resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/v2/KubernetesSparkDependencyService.scala

```diff
@@ -0,0 +1,106 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.deploy.rest.kubernetes.v2
+
+import java.io.InputStream
+import javax.ws.rs.{Consumes, GET, HeaderParam, Path, Produces, PUT, QueryParam}
+import javax.ws.rs.core.{MediaType, StreamingOutput}
+
+import org.glassfish.jersey.media.multipart.FormDataParam
+
+import org.apache.spark.deploy.rest.KubernetesCredentials
+
+/**
+ * Service that receives application data that can be received later on. This is primarily used
+ * in the context of Spark, but the concept is generic enough to be used for arbitrary applications.
+ * The use case is to have a place for Kubernetes application submitters to bootstrap dynamic,
+ * heavyweight application data for pods. Application submitters may have data stored on their
+ * local disks that they want to provide to the pods they create through the API server. ConfigMaps
+ * are one way to provide this data, but the data in ConfigMaps are stored in etcd which cannot
+ * maintain data in the hundreds of megabytes in size.<br>
+ * <br>
+ * The general use case is for an application submitter to ship the dependencies to the server via
+ * {@link uploadDependencies}; the application submitter will then receive a unique secure token.
+ * The application submitter then ought to convert the token into a secret, and use this secret in
+ * a pod that fetches the uploaded dependencies via {@link downloadJars}, {@link downloadFiles}, and
+ * {@link getKubernetesCredentials}.
+ */
+@Path("/")
+private[spark] trait KubernetesSparkDependencyService {
+
+  /**
+   * Register an application with the dependency service, so that the driver pod can retrieve them
+   * when it runs.
+   *
+   * @param driverPodName Name of the driver pod.
+   * @param driverPodNamespace Namespace for the driver pod.
+   * @param jars Application jars to upload, compacted together in tar + gzip format. The tarball
```

Comment:
The APIs here are worded in a Spark-opinionated way, but there could also be a world where we just take one tarball as input "files" and just have spark-submit call uploadFiles() twice - once for jars and once for local files. If we were to do this, we would need a different unique key for the application other than name and namespace indicating the pod(s) to watch, - probably a label(s).